### PR TITLE
Move the crate generator to cargo-progenitor

### PR DESCRIPTION
**Pull request by iliana**
The `progenitor` crate today has a library used by downstream client generators, and a binary for generating new client crates. The latter needs to know what dependencies Progenitor was built with so it can write out a correct Cargo.toml. It does this by [finding a Cargo.lock](https://docs.rs/project-root/0.2.2/src/project_root/lib.rs.html#19-34) using a build.rs script, and passing that to [built](https://docs.rs/built).

In a particularly contrived environment (i.e. nixpkgs, used for building stuff in NixOS), I've discovered this doesn't work. nixpkgs builds a `cargo vendor` tree and sticks the `progenitor` crate somewhere like `/build/vendor/progenitor-0.3.0`, and puts the source code of the project it's trying to build at somewhere like `/nix/store/sjdfkj8329kdcfj92dkfj-source`. In this environment, project-root fails to find the project Cargo.lock because it is not in an ancestor of progenitor's source directory, and the build.rs script fails.

But we don't need this build.rs to work to build the Progenitor _library_. We can take a page from [typify](https://github.com/oxidecomputer/typify) and make the crate generator a separate `cargo-progenitor` crate that only provides the crate generator binary. (This also happens to remove a bunch of dependencies from progenitor that downstream consumers probably don't need.)

### CODE REVIEW

File: cargo-progenitor/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,26 @@
+[package]
+name = "cargo-progenitor"
+version = "0.3.0"
+edition = "2021"
+license = "MPL-2.0"
+description = "An OpenAPI client generator"
```
Comment (ahl) [inline cargo-progenitor/Cargo.toml:6]:
```suggestion
description = "A cargo command to generate a Rust client SDK from OpenAPI"
```

File: cargo-progenitor/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+repository = "https://github.com/oxidecomputer/progenitor.git"
+readme = "../README.md"
```
Comment (ahl) [inline cargo-progenitor/Cargo.toml:8]: Probably deserves its own README, but that can wait for another day..

File: cargo-progenitor/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+keywords = ["openapi", "openapiv3", "sdk", "generator", "proc_macro"]
+categories = ["api-bindings", "compilers"]
```
Comment (ahl) [inline cargo-progenitor/Cargo.toml:10]:
```suggestion
categories = ["api-bindings", "development-tools::cargo-plugins"]
```

File: cargo-progenitor/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,26 @@
+[package]
+name = "cargo-progenitor"
+version = "0.3.0"
+edition = "2021"
+license = "MPL-2.0"
+description = "An OpenAPI client generator"
+repository = "https://github.com/oxidecomputer/progenitor.git"
+readme = "../README.md"
+keywords = ["openapi", "openapiv3", "sdk", "generator", "proc_macro"]
```
Comment (ahl) [inline cargo-progenitor/Cargo.toml:9]:
```suggestion
keywords = ["openapi", "openapiv3", "sdk", "generator"]
```

File: cargo-progenitor/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+categories = ["api-bindings", "compilers"]
+build = "build.rs"
+
+[dependencies]
+progenitor = { version = "0.3.0", path = "../progenitor" }
+progenitor-client = { version = "0.3.0", path = "../progenitor-client" }
```
Comment (ahl) [inline cargo-progenitor/Cargo.toml:15]: is this needed?

File: cargo-progenitor/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+progenitor-impl = { version = "0.3.0", path = "../progenitor-impl" }
+anyhow = "1.0"
+clap = { version = "4.2.7", features = ["derive"] }
+openapiv3 = "1.0.0"
```
Comment (ahl) [inline cargo-progenitor/Cargo.toml:19]: weird this didn't update...
```suggestion
openapiv3 = "1.0.2"
```

**Review by ahl — COMMENTED** [review]
looks good; can you make a minor update here:
https://github.com/oxidecomputer/progenitor/blob/main/README.md#static-crate

Maybe explaining that you can `cargo install cargo-progenitor` or `cargo run progenitor ...`?

File: cargo-progenitor/Cargo.toml (hunk 6 above)
Comment (iliana) [reply cargo-progenitor/Cargo.toml:19]: I used copy/paste instead of `cargo add` :)

File: cargo-progenitor/Cargo.toml (hunk 5 above)
Comment (iliana) [reply cargo-progenitor/Cargo.toml:15]:
```
   Compiling cargo-progenitor v0.3.0 (/Users/iliana/git/progenitor/cargo-progenitor)
error[E0433]: failed to resolve: use of undeclared crate or module `progenitor_client`
   --> cargo-progenitor/src/main.rs:210:46
    |
210 |                 let progenitor_client_code = progenitor_client::code();
    |                                              ^^^^^^^^^^^^^^^^^ use of undeclared crate or module `progenitor_client`
    |
help: there is a crate or module with a similar name
    |
210 |                 let progenitor_client_code = progenitor_impl::code();
    |                                              ~~~~~~~~~~~~~~~

For more information about this error, try `rustc --explain E0433`.
error: could not compile `cargo-progenitor` due to previous error
```

**Comment by iliana** [comment]
One confusing thing about cargo plugins is that if you want to use `cargo run`, you need to pass in the extra `progenitor` parameter (e.g. `cargo run --bin cargo-progenitor -- progenitor --whatever`). Not sure how I wanted to go about explaining this in the README so I didn't.

It might make more sense to be smarter about parsing arguments to support either `cargo progenitor --whatever` or `cargo-progenitor --whatever` being passed in as argv.

File: cargo-progenitor/Cargo.toml (hunk 5 above)
Comment (ahl) [reply cargo-progenitor/Cargo.toml:15]: indeed it is!

**Comment by ahl** [comment]
> One confusing thing about cargo plugins is that if you want to use `cargo run`, you need to pass in the extra `progenitor` parameter (e.g. `cargo run --bin cargo-progenitor -- progenitor --whatever`). Not sure how I wanted to go about explaining this in the README so I didn't.
> 
> It might make more sense to be smarter about parsing arguments to support either `cargo progenitor --whatever` or `cargo-progenitor --whatever` being passed in as argv.

I think: meh. I'd like to nudge people towards doing the `cargo install` anyways.

File: README.md [hunk 7]
```diff
@@ -161,7 +161,10 @@ Options:
 
 For example:
 
-`cargo run --bin progenitor -- -i sample_openapi/keeper.json -o keeper -n keeper -v 0.1.0`
+```
+cargo install --path cargo-progenitor
+cargo progenitor -i sample_openapi/keeper.json -o keeper -n keeper -v 0.1.0
+```
```
Comment (ahl) [inline README.md:167]:
``````suggestion
```
cargo install cargo-progenitor
cargo progenitor -i sample_openapi/keeper.json -o keeper -n keeper -v 0.1.0
```

... or within the repo:
```
cargo run --bin cargo-progenitor -- progenitor -i sample_openapi/keeper.json -o keeper -n keeper -v 0.1.0
```
``````

**Review by ahl — APPROVED** [review]
one minor suggestion; take it or leave it

**Pull request merged**